Libs/Visualization/VTK/Widgets/ctkVTKAbstractView.cpp

```diff
@@ -70,7 +70,11 @@ void ctkVTKAbstractViewPrivate::init()
   this->setParent(q);
 
 #if CTK_USE_QVTKOPENGLWIDGET
+# if CTK_HAS_QVTKOPENGLNATIVEWIDGET_H
```

Comment:
It is not ideal to have this repeated code scattered around in the code base. It makes code less readable and if VTK changes then we have to touch many places in the code again.

Instead, we could generate a header file, such as ctkVTKNativeWidget.h, which would have something like this inside:

```
#if CTK_USE_QVTKOPENGLWIDGET
# if CTK_HAS_QVTKOPENGLNATIVEWIDGET_H
#  include <QVTKOpenGLNativeWidget.h>
# else
#  include <QVTKOpenGLWidget.h>
# endif
#else
#include <QVTKWidget.h>
#endif

#if CTK_USE_QVTKOPENGLWIDGET
# if CTK_HAS_QVTKOPENGLNATIVEWIDGET_H
typedef ctkVTKOpenGLNativeWidget QVTKOpenGLNativeWidget;
# else
typedef ctkVTKOpenGLNativeWidget QVTKOpenGLWidget;
# endif
#else
typedef ctkVTKOpenGLNativeWidget QVTKWidget;
#endif
```

It would be enough to include a single file and use a single type everywhere else in CTK (and code that uses CTK) to have access to OpenGL native widget in a backward compatible manner.

Reply:
That would word but I fear this would obfuscate the name of the actual widget in the class and make it harder to understand what is what ... . Indeed their API is different .. and seeing the same class name used with different API would be quite strange.

Reply:
Since I tested these changes ... I would like to move forward with this.

Reply:
Thanks a lot for working on this. Using latest VTK master with all the fixes and improvements will be awesome!

Reply:
I wonder if PythonQt will still behave ..

Reply:
Currently works .. I will test and report back:

```
>>> slicer.app.layoutManager().threeDWidget(0).threeDView().VTKWidget()
QVTKOpenGLNativeWidget(0x5b309d0) 
```

Reply:
I guess it would, but if not then we could try to use #define instead of typedef.

Reply:
As side note, as soon as we are done with this ... I will have a PR ready for Slicer.

Reply:
Great! I would like to test all new renderer features and fixes but could not without these CTK and Slicer fixes.